test: sensor : add module testcase of sensor

yingmingx opened the pull request:
Add module test of API sensor_sample_fetch_chan().
    Add module test of unit coversion function.


Signed-off-by: Ying ming <mingx.ying@intel.com>

LeiW000 commented on tests/drivers/sensor/generic/src/main.c:
```diff
@@ -111,11 +111,22 @@ void test_sensor_get_channels(void)
 
 	/* Get and check channels value. */
-	for (int i = 0; i < TOTAL_CHAN_ELEMENTS; i++) {
+	for (int i = 1; i < TOTAL_CHAN_ELEMENTS; i++) {
```
We don't need to run testing for channel 0 any more?

yingmingx replied:
Sensor data in chanel 0 is taken into ram by the supplement api sensor_sample_fetch_chan() and obtained by cpu use sensor_channel_get() at the begin of this testcase. 
So I think the date in chanel 0 is no need to be get by cpu twice.

enjiamai commented on tests/drivers/sensor/generic/src/main.c:
```diff
@@ -111,11 +111,22 @@ void test_sensor_get_channels(void)
+	zassert_equal(data.val1, chan_elements[0].data.val1,
+				"the data is not match.");
+	zassert_equal(data.val2, chan_elements[0].data.val2,
+				"the data is not match.");
```
Excuse me, don't know understand why don't you put line 114~123 inside the loop together? They seems doing the same action but different channel.

yingmingx replied:
The purpose of  this PR is to invoke the api sensor_sample_fetch_chan() which didn't test in previous code.  The difference between sensor_sample_fetch() and sensor_sample_fetch_chan() is the number of channel will be read.  sensor_sample_fetch() read all the channel and sensor_sample_fetch_chan() read only one chanel. The user may then get individual channel values by calling sensor_channel_get. 
I used sensor_sample_fetch_chan() to read chanel 0 and use sensor_channel_get to get the chanel vaule.
Then, read all the channel by sensor_sample_fetch and get the remaining chanel vaule from 1-TOTAL_CHAN_ELEMENTS by sensor_channel_get .
After that, all the chanel value will be get only once.

enjiamai replied:
It looks weird from my point of view.  Here is some suggestions, just for you reference.

static void check_channel_value()
{
        struct sensor_value data;

        zassert_equal(data.val1, chan_elements[i].data.val1,
        		"the data is not match.");
        zassert_equal(data.val2, chan_elements[i].data.val2,
                        "the data is not match.");
}

void test_sensor_get_channels(void)
{
        const struct device *dev;

        dev = device_get_binding(DUMMY_SENSOR_NAME);
        zassert_not_null(dev, "failed: dev is null.");

	/* test fetch all channel */
        zassert_equal(sensor_sample_fetch(dev), RETURN_SUCCESS,
                        "fail to fetch sample.");

        /* Get and check channels value. */
        for (int i = 0; i < TOTAL_CHAN_ELEMENTS; i++) {
                zassert_equal(sensor_channel_get(dev, chan_elements[i].chan,
                                &data), RETURN_SUCCESS, "fail to get channel.");

		check_channel_vaule();

		/* test fetch single channel */
		zassert_equal(sensor_sample_fetch_chan(dev, chan_elements[0].chan),
				RETURN_SUCCESS,	"fail to fetch sample.");

		check_channel_vaule();
        }
}

yingmingx replied:
Thank you for your sample code. In this way, the code can run.  
Strictly speaking，your code didn't show the api sensor_sample_fetch_chan() could fetch the channel value because the api sensor_sample_fetch() outside the loop has fetch all the channel data.
Althought my code seems weird. But I use sensor_sample_fetch_chan at the begin of the testcase. This means it can works without rely on the api sensor_sample_fetch().

enjiamai replied:
I understand the logic, maybe split the up would be better idea. It's not mandatory from my side.

enjiamai reviewed (changes requested):
The logic seems duplicated to me, could you please explain that?

Commit:
test: sensor: add testcase

Add module test of API sensor_sample_fetch_chan().
Add module test of unit coversion function.

Signed-off-by: Ying ming <mingx.ying@intel.com>

LeiW000 commented:
@nashif , could you help to merge it?

enjiamai reviewed (approved):
No problem from my side, thanks.
